lib/fibonacci.rb

```diff
@@ -4,6 +4,39 @@
 # e.g. 1st fibonacci number is 1
 # ....
 # e.g. 6th fibonacci number is 8
+
+#Time complexity is O(n) because it linearly increases based on the size of the input number
+#Space complexity is O(1) because the number of variables need is constant regardless of the size of input number
 def fibonacci(n)
-  raise NotImplementedError
+  raise ArgumentError if n == nil
+  raise ArgumentError if n < 0
+
+  i = 0
+  k = 0
+  m = 0
+  j = 1
+
+  until i > n
+    k = k + m
+    m = j
+    j = k
+    i += 1
+  end
+
+  return k
 end
+
+# def fibonacci(n)
```

Comment:
Nice work thinking about this recursively. It turns out that the space complexity of the recursive approach to fibonacci, like in your algorithm results in a time complexity of O(2^n). You already know that exponential is one of the worst time complexity. Also, the space complexity of this algorithm will be O(n) due to the call stack space it uses. We'll be looking to understand the time and space complexity of this approach more closely later in class.